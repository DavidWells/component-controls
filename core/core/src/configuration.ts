import { Configuration as WebpackConfiguration } from 'webpack';
import { ActionItems } from '@component-controls/components';
import { ComponentType, ReactNode } from 'react';
import { StoryRenderFn } from './utility';

/**
 * story type pages can have multiple tabs with separate page configurations.
 */
export interface TabConfiguration {
  /**
   * tab route string
   */
  route?: string;
  /**
   * title will be used as tab caption
   */
  title: string;
  /**
   * page container type - a key into the component-controls/pages package
   */
  type?: string;
  /**
   * render function, returns a react component
   */
  render?: (props: any) => ReactNode;
}

export type PageTabs = TabConfiguration[];

export type DocumentType =
  | 'story'
  | 'blog'
  | 'page'
  | 'tags'
  | 'author'
  | string;

export interface PageConfiguration {
  /**
   * base url path for the page
   */
  basePath?: string;

  /**
   * if true, generate story-based paths. This is for documents with a navSidebar that would allow selection of specific stories.
   */
  storyPaths?: boolean;

  /**
   * label - used for menu labels
   */
  label?: string;

  /**
   * whether to take a fullpage theme option
   */
  fullPage?: boolean;

  /**
   * whether to have an index home page for the doc type.
   * if false, will show the first document of the doc type as the home page.
   */
  indexHome?: boolean;

  /**
   * whether to add navigation sidebar to the page
   */
  navSidebar?: boolean;

  /**
   * whether to add conext sidebar to navigate the sections of the current document
   */
  contextSidebar?: boolean;

  /**
   * whether to add to the top navigation menu
   */
  topMenu?: boolean;

  /**
   * page container react component
   */
  container?: ComponentType | null;

  /**
   * tabs configuration for story-type pages
   */
  tabs?: PageTabs;
}

export type PagesConfiguration = Record<DocumentType, PageConfiguration>;

type WebpackConfigFn = (
  config: WebpackConfiguration,
  options?: any,
) => WebpackConfiguration;
type WebpackConfig = WebpackConfiguration | WebpackConfigFn;

export type PagesOnlyRoutes = Record<
  DocumentType,
  Pick<PageConfiguration, 'basePath' | 'storyPaths'> & {
    tabs?: Pick<TabConfiguration, 'route'>[];
  }
>;
/**
 * global configuration used at build time
 * stored in a file named main.js/main.ts
 */
export interface BuildConfiguration {
  /**
   * wild card search string for the stories
   * internally using `glob` for the search: https://www.npmjs.com/package/glob
   * example: "./stories/**/ /*.stories.(js|jsx|tsx|mdx)"
   */
  stories?: string[];
  /**
   * base url path for API documentation pages. Default is "docs/"
   */
  pages?: PagesOnlyRoutes;

  /**
   * page types that are considered as categories fields as well
   */
  categories?: DocumentType[];
  /**
   * custom webpack configuration setup. One or the other will be used
   */
  webpack?: WebpackConfig;
  finalWebpack?: WebpackConfig;
}

export interface ToolbarConfig {
  /**
   * left side toolbar items
   */
  left?: ActionItems;

  /**
   * right side toolbar items
   */
  right?: ActionItems;
}
/**
 * global configuration used at build time
 * stored in a file named main.js/main.ts
 */
export interface RunOnlyConfiguration {
  /**
   * story decorator functions - used to wrap stories
   * example: [story => <ThemeProvider>{story()}</ThemeProvider>]
   */
  decorators?: StoryRenderFn[];
  /**
   * standalone site title. Default is "Component controls"
   */
  siteTitle?: string;
  /**
   * site alt for images. Default is "Component controls - https://github.com/ccontrols/component-controls"
   */
  siteTitleAlt?: string;

  /**
   * Site headline. Default is "Component controls gatsby"
   */
  siteHeadline?: string;

  /**
   * Deployed site url. Default is "https://component-controls-gatsby.netlify.app"
   */
  siteUrl?: string;

  /**
   * site description. siteDescription: Default is "Component controls stories. Write your components documentation with MDX and JSX. Design, develop, test and review in a single site."
   */
  siteDescription?: string;

  /**
   * site language, Deault is "en"
   */
  siteLanguage?: string;

  /**
   * author: Default is "@component-controls"
   */
  author?: string;

  /**
   * link to site image
   */
  siteImage?: string;

  /**
   * page types configurations
   */
  pages?: PagesConfiguration;

  /**
   * theme-ui theme configuration
   */
  theme?: { [key: string]: any };

  /**
   * story sorting function
   */
  storySort?: (a: string, b: string) => number;

  /**
   * custom toolbar items
   */
  toolbar?: ToolbarConfig;
}

export type RunConfiguration = RunOnlyConfiguration &
  Omit<BuildConfiguration, 'pages'>;

export const defaultRunConfig: RunConfiguration = {
  siteTitle: 'Component controls',
  siteTitleAlt:
    'Component controls - https://github.com/ccontrols/component-controls',
  siteHeadline: 'Component controls gatsby',
  siteUrl: 'https://component-controls-gatsby.netlify.app',
  siteDescription:
    'Component controls stories. Write your components documentation with MDX and JSX. Design, develop, test and review in a single site.',
  siteLanguage: 'en',
  author: '@component-controls',
  pages: {
    story: {
      label: 'Docs',
      navSidebar: true,
      contextSidebar: true,
      topMenu: true,
      tabs: [
        { title: 'Documentation', type: 'ClassicPage' },
        { title: 'Testing', type: 'TestingPage' },
      ],
    },
    blog: {
      label: 'Blog',
      contextSidebar: true,
      topMenu: true,
      indexHome: true,
    },
    author: {
      label: 'Authors',
    },
    page: {
      label: 'Page',
      container: null,
    },
    tags: {
      label: 'Tags',
    },
  },
};

export const defaultBuildConfig: BuildConfiguration = {
  categories: ['author', 'tags'],
  pages: {
    story: {
      basePath: 'docs/',
      storyPaths: true,
      tabs: [{ route: 'page' }, { route: 'test' }],
    },
    blog: {
      basePath: 'blogs/',
    },
    author: {
      basePath: 'authors/',
    },
    page: {
      basePath: 'pages/',
    },
    tags: {
      basePath: 'tags/',
    },
  },
};
