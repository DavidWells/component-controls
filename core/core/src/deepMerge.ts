import merge from 'deepmerge';

const concatMerge = (dest: any[], src: any[]) => [...dest, ...src];

export const deepMerge = <T>(a: any, b: any) =>
  merge<T>(a, b, { arrayMerge: concatMerge });

const mergeArrays = (dest: any[], src: any[]) => {
  const result =
    src.length > 0
      ? src.map((s, idx) =>
          idx < dest.length ? deepMergeArrays<any[]>(dest[idx], s) : s,
        )
      : src;
  return result;
};
export const deepMergeArrays = <T>(a: any, b: any) =>
  merge<T>(a, b, { arrayMerge: mergeArrays });

export const deepMergeReplaceArrays = <T>(a: any, b: any) =>
  merge<T>(a, b, {
    arrayMerge: (dest: any[], src: any[]) => src,
  });

export { deepMergeReplaceArrays as merge };

export { merge as deepmerge };
