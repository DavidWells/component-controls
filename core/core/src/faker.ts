import * as faker from 'faker/locale/en_US';

export { faker };
