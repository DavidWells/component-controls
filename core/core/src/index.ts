export * from './utils';
export { randomizeData } from './randomizeData';
export * from './smartControls';
export * from './source';
export * from './faker';
