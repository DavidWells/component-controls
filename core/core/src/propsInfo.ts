import { ComponentInfo } from './components';

/**
 * callback function to extract props info table  - ie docgen type libraries
 * used to extract displayName, and props tables for a component
 *
 * @param fileName full name and path of the component path react-docgen needs it to extract babel configurations.
 * @param componentName optional component name react-docgen-typescript supports multiple exports for a file react-docgne does not use it.
 * @param source optional soure, saves time if its already loaded react-docgen accepts source as input parameter react-docgen-typescript does not use it.
 *
 * @returns async or normal function that loads ComponentInfo props info.
 */
export type PropsInfoExtractorFunction = (
  fileName: string,
  componentName?: string,
  source?: string,
) => Promise<ComponentInfo | undefined> | ComponentInfo | undefined;
