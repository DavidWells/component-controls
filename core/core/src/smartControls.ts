/* eslint-disable no-console */
import {
  ComponentControl,
  ComponentControls,
  ControlTypes,
} from '@component-controls/specification';

export interface PropType {
  defaultValue: { value?: string | boolean; computed: boolean };
  type: { name?: string };
  required: boolean;
  description: string;
}

export interface PropTypes {
  [key: string]: PropType;
}

const cleanQuotes = (txt?: string) => (txt ? txt.replace(/['"]+/g, '') : txt);

const handledTypes = [
  'boolean',
  'bool',
  'string',
  'number',
  'enum',
  'func',
  'shape',
  '(() => void)',
];

export const controlFromProps = (
  name: string,
  propDef: PropType,
): ComponentControl | null => {
  if (!propDef) {
    return null;
  }
  let type = propDef.type.name;
  let splitType: string[] | undefined = undefined;
  // docgen typescript are ie "boolean | undefined"
  if (typeof type === 'string') {
    splitType = type.split(' | ');
    if (splitType.length > 1) {
      let found: string | undefined;
      // we have a typescrpit type definitio of "|" type
      for (let i = 0; i < splitType.length; i += 1) {
        const match = splitType[i];
        found = handledTypes.find(a => a === match);
        if (found !== undefined) {
          type = found;
          break;
        }
      }
      if (found === undefined) {
        type = 'enum';
      }
    }
  }
  function onClick() {
    // eslint-disable-next-line prefer-rest-params
    console.info(`${name}: `, arguments);
  }
  const defaultValue = propDef.defaultValue
    ? propDef.defaultValue.value
    : undefined;
  switch (type) {
    case 'string': {
      let value: string | undefined;
      if (typeof defaultValue === 'string') {
        value = defaultValue;
      }
      value = cleanQuotes(value);
      const isColor = name.toLowerCase().includes('color');
      if (!value && propDef.required) {
        value = isColor ? 'red' : 'example';
      }
      return {
        type: isColor ? ControlTypes.COLOR : ControlTypes.TEXT,
        value,
      };
    }
    case 'boolean':
    case 'bool': {
      let value;
      if (defaultValue !== undefined) {
        // docgen typescript defaultValue.summary is actually a boolean type, not a string
        if (defaultValue === 'false' || (defaultValue as unknown) === false) {
          value = false;
        }
        if (defaultValue === 'true' || (defaultValue as unknown) === true) {
          value = true;
        }
      }
      return { type: ControlTypes.BOOLEAN, value };
    }
    case 'number': {
      let value;
      try {
        if (typeof defaultValue === 'string') {
          value = parseFloat(defaultValue);
        }
      } catch (e) {
        // eat exceptoin
      }
      return { type: ControlTypes.NUMBER, value };
    }
    case 'enum': {
      const value =
        typeof defaultValue === 'string'
          ? cleanQuotes(defaultValue)
          : undefined;
      const options = Array.isArray(propDef.type)
        ? propDef.type
        : splitType || (propDef.type as any).value;
      if (!Array.isArray(options)) {
        return null;
      }
      return {
        type: ControlTypes.OPTIONS,
        options: options.map((v: any) => cleanQuotes(v.value ? v.value : v)),
        value,
      };
    }
    // typescript callback function signature
    case '(() => void)':
    case 'func': {
      return {
        type: ControlTypes.BUTTON,
        label: name,
        onClick,
        value: onClick,
      };
    }
    case 'shape': {
      /* let value;
      try {
        if (propDef.defaultValue && typeof propDef.defaultValue.summary === 'object') {
          value = JSON.parse(propDef.defaultValue.summary);
        }
      } catch (e) {
        // eat exception
      }
      */
      return null;
    }
    default:
      return null;
  }
};

interface NamedComponentControl {
  name: string;
  control: ComponentControl | null;
}
export const controlsFromProps = (props: PropTypes): ComponentControls => {
  return Object.keys(props)
    .map((key: string) => ({
      name: key,
      control: controlFromProps(key, props[key]),
    }))
    .filter(p => p.control)
    .reduce(
      (acc: ComponentControls, prop: NamedComponentControl) => ({
        ...acc,
        [prop.name]: prop.control as any,
      }),
      {},
    );
};
