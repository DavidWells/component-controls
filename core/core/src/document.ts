import { CodeLocation, PackageInfo, StoryRenderFn } from './utility';
import { StoryComponent } from './components';
import { ComponentControls } from './controls';
import { RunConfiguration, DocType, PageLayout } from './configuration';
/**
 * an identifier/variable.argument in the source code
 */
export interface SourceIdentifier {
  name: string;
  loc?: CodeLocation;
}
export interface ArgUsageLocation {
  /**
   * where in the story source code is the argument used
   * code location is relative to the start of the story
   */
  loc: CodeLocation;

  /**
   * optional name for the usage of the argument
   * example: export const story = ({ value }) => <Story value={{ age: value }} />;
   * in this example the name will be 'age'
   */
  name?: SourceIdentifier;

  /** true if the property is a 'shorthand'.
   * { prop: value } - not a shorthand.
   * { prop } - a shorthand.
   */
  shorthand?: boolean;
}

/**
 * arguments passed to the 'story' function, extracted by an AST loader
 */
export interface StoryArgument {
  /**
   * either the name used (or a variable alias), or an array of sub-arguments
   * ({ name: alias })
   */
  value: string | StoryArguments;
  /**
   * the original name of the argument
   */
  name?: string;
  /**
   * location of the argument declaration, relative to the story source code
   */
  loc?: CodeLocation;
  /**
   * list of locations where the argument is used in the body of the story
   */
  usage?: ArgUsageLocation[];
}

/**
 * list of story arguments. Each argument can be a deconstructed argument of itself
 * the first argument are the control 'values'
 */
export type StoryArguments = StoryArgument[];

export interface SmartControls {
  /**
   * whether to generate "smart" controls for a story
   */
  smart?: boolean;
  /**
   * include props only
   */
  include?: string[];

  /**
   * exclude props only
   */
  exclude?: string[];
}

/**
 * story prooperties that can be inherited from the document, or each story can have its properties
 */

export interface StoryProps {
  /**
   * id for component associated with the story
   */
  component?: string | object;

  /**
   * multiple components option
   */
  subcomponents?: {
    [key: string]: string | object;
  };

  /**
   * object of key/value pairs specifying the controls for the story
   */
  controls?: ComponentControls;

  /**
   * "smart" controls options
   */
  smartControls?: SmartControls;
  /**
   * array of wrapper functions (decorators) to be called when rendering each individual story.
   */
  decorators?: StoryRenderFn[];
}

/**
 * Story interface - usually extracted by the AST instrumenting loader
 */
export type Story = {
  /**
   * name of the Story.
   */
  name: string;

  /**
   * id of the story
   */
  id?: string;

  /**
   * title of the file/group of stories
   */
  doc?: string;

  /**
   * render function for the story
   */
  renderFn?: StoryRenderFn;

  /**
   * story extended description. can use markdown.
   */
  description?: string;

  /**
   * arguments passed to the story function.
   * eg `export const story = props => <Story {...props} />;`
   */
  arguments?: StoryArguments;
  /**
   * location in the source file of the story definition
   */
  loc?: CodeLocation;
  /**
   * the source code of the story, extracted by the AST instrumenting loaders
   */
  source?: string;

  /**
   * optional story subtitle property
   */
  subtitle?: string;
} & StoryProps;

/**
 * map of stories. The id is compatible with storybook's story ids
 */
export interface Stories {
  [id: string]: Story;
}

export const defDocType: DocType = 'story';
/**
 * A documentation file's metadata.
 * For MDX files, fromtmatter is used to declare the document properties.
 * For ESM (ES Modules) documentation files, the default export is used.
 */
export type Document = {
  /**
   * title of the document. If no 'route' parameter is specifified, the title is used to generate the document url.
   * This is the only required field, to show the document in the menu structures.
   */
  title: string;

  /**
   * document type - blogs, pages, stories and even custom ones. By default - story
   */
  type?: DocType;

  /**
   * if provided, will be used as the route for the page.
   * if not provided, the title in lowercase will be used as the route
   */
  route?: string;

  /**
   * page layout - sidebars, full width
   */
  layout?: PageLayout;

  /**
   *  optional date the document was created. If not assigned, the instrumentation process will use birthtime
   */
  date?: Date;

  /**
   *  optional date the document was last modified. If not assigned, the instrumentation process will use mtime
   */
  dateModified?: Date;

  /**
   * if set to true, the document will be hidden in production builds.
   */
  draft?: boolean;

  /**
   *  comma-separated list of document tags, used for search
   */
  tags?: string[];

  /**
   *  documentation file description
   */
  description?: string;

  /**
   * document author
   */
  author?: string;

  /**
   * document order, used to sort documents within the same parent
   */
  order?: number;

  //next generated by the instrumentation process

  /**
   * list of story ids contained in the document.
   */
  stories?: string[];

  /**
   * source code of the entire file of stories
   */
  source?: string;

  /**
   * file name of the file of stories
   */
  fileName?: string;

  /**
   * lookup into the global store of PackageInfo package.json
   */
  package?: string;

  /**
   * lookup into the global store.components
   * since multiple components of the same name can be used
   * example: ['Button']: 'c:/myapp/Button.tsx'
   */
  componentsLookup: {
    [name: string]: string;
  };

  /**
   * for MDX documents, this is an MDXContent function, to be rendered inside a MDXProvider
   */
  MDXPage?: any;

  /**
   * custom prop set by mdxjs
   */
  isMDXComponent?: boolean;
} & StoryProps;

export const dateToLocalString = (date?: Date): string =>
  date
    ? new Date(date).toLocaleDateString('en-US', {
        timeZone: 'UTC',
      })
    : '';
/**
 * list of components used in stories
 */

export interface StoreComponents {
  [fileName: string]: StoryComponent;
}

/**
 * list of story files, or groups
 */
export interface Documents {
  [title: string]: Document;
}

export type Pages = Document[];

/**
 * list of stories
 */
export interface StoreStories {
  [id: string]: Story;
}

/**
 * list of repositories
 */
export interface StorePackages {
  [id: string]: PackageInfo;
}

export type StoreObserver = (story?: Story) => void;

export const CURRENT_STORY = '.';
/**
 * store of stories information in memory after the loader is applied
 */
export interface Store {
  /**
   * global configuration for config file
   */
  config?: RunConfiguration;
  /**
   * list of story files, or groups
   */
  docs: Documents;
  /**
   * list of stories
   */
  stories: StoreStories;
  /**
   * list of components used in stories
   */
  components: StoreComponents;

  /**
   * list of package.json files and their data
   * used by the components and the stories of the project
   */
  packages: StorePackages;

  /**
   * storybook integration notifiers
   */
  addObserver: (observer: StoreObserver) => void;
  removeObserver: (observer: StoreObserver) => void;
  /**
   * update store, for example controls or state
   */
  updateStory: (story: Story) => void;
}

export const defaultStore: Store = {
  components: {},
  docs: {},
  packages: {},
  stories: {},
  addObserver: () => {},
  removeObserver: () => {},
  updateStory: function(story: Story) {
    if (story) {
      if (story.id) {
        this.stories = {
          ...this.stories,
          [story.id]: story,
        };
      }
    }
  },
};
