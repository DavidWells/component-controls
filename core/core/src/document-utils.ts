import {
  toId,
  storyNameFromExport as csfStoryNameFromExport,
} from '@storybook/csf';
import { PagesOnlyRoutes, DocType, PageConfiguration } from './configuration';
import { Document, defDocType } from './document';

export const storyNameFromExport = csfStoryNameFromExport;
export const strToId = (str: string) => str.replace(/\W/g, '-').toLowerCase();

export const ensureTrailingSlash = (route: string) =>
  route.endsWith('/') ? route : route + '/';

export const ensureStartingSlash = (route: string) =>
  route.startsWith('/') ? route : '/' + route;

export const removeTrailingSlash = (route: string) =>
  route.endsWith('/') ? route.substr(0, route.length - 1) : route;

export const getDocPath = (
  docType: DocType,
  doc?: Document,
  pagesConfig?: PagesOnlyRoutes,
  name: string = '',
  activeTab?: string,
): string => {
  const { basePath = '', storyPaths } = pagesConfig?.[docType] || {};
  if (storyPaths && doc && doc.stories && doc.stories.length > 0) {
    return getStoryPath(doc.stories[0], doc, pagesConfig, activeTab);
  }
  const route = doc
    ? doc.route ||
      `${ensureStartingSlash(
        ensureTrailingSlash(basePath),
      )}${ensureTrailingSlash(strToId(doc.title))}${
        activeTab ? `${ensureTrailingSlash(activeTab)}` : ''
      }`
    : `${ensureStartingSlash(
        ensureTrailingSlash(basePath),
      )}${ensureTrailingSlash(strToId(name))}${
        activeTab ? `${ensureTrailingSlash(activeTab)}` : ''
      }`;
  return removeTrailingSlash(route);
};

export const getStoryPath = (
  storyId?: string,
  doc?: Document,
  pagesConfig?: PagesOnlyRoutes,
  activeTab?: string,
): string => {
  const docType = doc?.type || defDocType;
  if (!storyId) {
    return getDocPath(docType, doc, pagesConfig, undefined, activeTab);
  }

  const { basePath = '' } = pagesConfig?.[docType] || {};
  const docRoute = `${
    doc?.route
      ? ensureStartingSlash(ensureTrailingSlash(doc?.route))
      : `${ensureStartingSlash(ensureTrailingSlash(basePath))}`
  }`;
  const route = `${docRoute}${storyId ? ensureTrailingSlash(storyId) : ''}${
    activeTab ? `${ensureTrailingSlash(activeTab)}` : ''
  }`;
  return removeTrailingSlash(route);
};

export const getDocTypePath = (type: PageConfiguration) =>
  type.basePath
    ? removeTrailingSlash(ensureStartingSlash(type.basePath))
    : undefined;

export const docStoryToId = (docId: string, storyId: string) =>
  toId(docId, storyNameFromExport(storyId));
