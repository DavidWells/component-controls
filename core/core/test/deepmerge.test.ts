import { deepMerge, deepMergeArrays } from '../src';
describe('deep merge', () => {
  it('Should append arrays', () => {
    expect(
      deepMerge(
        {
          a: [{ a: 'a', b: 'b' }],
        },
        {
          a: [{ c: 'c', d: 'd' }],
        },
      ),
    ).toMatchSnapshot();
  });

  it('Should merge arrays', () => {
    expect(
      deepMergeArrays(
        {
          a: [{ a: 'a', b: 'b' }],
        },
        {
          a: [{ c: 'c', d: 'd' }],
        },
      ),
    ).toMatchSnapshot();
  });
  it('Should merge arrays and fields', () => {
    expect(
      deepMergeArrays(
        {
          a: [
            { a: 'a', b: 'b' },
            { c: 'c', d: 'd' },
          ],
        },
        {
          a: [{ b: 'c' }],
        },
      ),
    ).toMatchSnapshot();
  });
  it('Should merge arrays mistmatched fields', () => {
    expect(
      deepMergeArrays(
        {
          b: [{ a: 'a', b: 'b' }],
        },
        {
          a: [{ b: 'c' }],
        },
      ),
    ).toMatchSnapshot();
  });
});
