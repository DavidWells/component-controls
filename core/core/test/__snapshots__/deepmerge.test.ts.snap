// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deep merge Should append arrays 1`] = `
Object {
  "a": Array [
    Object {
      "a": "a",
      "b": "b",
    },
    Object {
      "c": "c",
      "d": "d",
    },
  ],
}
`;

exports[`deep merge Should merge arrays 1`] = `
Object {
  "a": Array [
    Object {
      "a": "a",
      "b": "b",
      "c": "c",
      "d": "d",
    },
  ],
}
`;

exports[`deep merge Should merge arrays and fields 1`] = `
Object {
  "a": Array [
    Object {
      "a": "a",
      "b": "c",
    },
  ],
}
`;

exports[`deep merge Should merge arrays mistmatched fields 1`] = `
Object {
  "a": Array [
    Object {
      "b": "c",
    },
  ],
  "b": Array [
    Object {
      "a": "a",
      "b": "b",
    },
  ],
}
`;
